src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,291 @@
+import duke.tasks.Task;
+import duke.tasks.TaskDeadline;
+import duke.tasks.TaskEvent;
+import duke.tasks.TaskList;
+import duke.tasks.TaskTodo;
+
+/**
+ * A simple task list program.
+ * 
+ * @see https://github.com/nus-cs2103-AY2223S2/ip
+ */
+public class Duke {
+    private Ui ui = new Ui();
+    private Storage storage = new Storage("_duke_data.txt");
+    private TaskList tasks;
+    private Parser currentCommand;
+
+    public Duke() {
+        // Attempt to load task list from save file.
+        try {
+            this.tasks = this.storage.load();
+        } catch (DukeSaveLoadException e) {
+            this.ui.showError(e);
+            this.tasks = new TaskList();
+        }
+    }
+
+    /**
+     * Runs the app.
+     */
+    public void run() {
+        String logo = " ____        _        \n"
+                + "|  _ \\ _   _| | _____ \n"
+                + "| | | | | | | |/ / _ \\\n"
+                + "| |_| | |_| |   <  __/\n"
+                + "|____/ \\__,_|_|\\_\\___|\n";
+        this.ui.show("Hello from\n" + logo);
+
+        whileLoop:
+        while (true) {
+            if (!this.ui.hasCommand()) {
+                continue;
+            }
+            
+            String input = this.ui.readCommand();
+            this.currentCommand = new Parser(input);
+        
+            try {
+                switch (this.currentCommand.baseCommand) {
+                case "todo":
+                    this.addTodo();
+                    break;
+                case "deadline":
+                    this.addDeadline();
+                    break;
+                case "event":
+                    this.addEvent();
+                    break;
+                case "mark":
+                    this.mark();
+                    break;
+                case "unmark":
+                    this.unmark();
+                    break;
+                case "delete":
+                    this.delete();
+                    break;
+                case "list":
+                    this.list();
+                    break;
+                case "find":
+                    this.find();
+                    break;
+                case "q":
+                case "quit":
+                case "exit":
+                case "bye":
+                    this.ui.show("Bye. Hope to see you again soon!");
+                    break whileLoop;
+                default:
+                    throw new DukeInvalidCommandException();
+                }
+
+                this.storage.save(this.tasks);
+            } catch (DukeException e) {
+                this.ui.showError(e);
+            }
+        }
+
+        this.ui.close();
+    }
+
+    public static void main(String[] args) {
+        new Duke().run();
+    }
+
+    /**
+     * Handles the adding of todo-type tasks by the "todo" command.
+     * 
+     * @throws DukeInvalidArgumentException If no description given.
+     */
+    private void addTodo() throws DukeInvalidArgumentException {
```

Comment:
Maybe the implementation of the commands can be abstracted out as their own classes under a command super class?